ramips: add support for Joowin WR758AC V1 and V2

araujorm opened the pull request:
Signed-off-by: Rodrigo Araujo <araujo.rm@gmail.com>

This patch adds support for Joowin (aka Comfast) WR758AC V1 and V2 devices.

Both are MT7628AN based, but V1 has a MT7662 as its 5GHz chip, while V2 has a MT7663 as its 5GHz chip. Some units are apparently mislabelled as V1, so if when you boot OpenWRT you don't see the 5GHz WiFi chip in LuCI, try the other image.

To install, setup a TFTP server on a machine with IP address 192.168.1.10 connected to the device's ethernet port and rename the squashfs-sysupgrade.bin to firmware_auto.bin. Turn off the device using the physical switch, wait a few seconds and turn it on again. The device should pull the firmware_auto.bin from the TFTP server, when it does disconnect the cable or stop the TFTP server, and the device should boot up with OpenWRT installed and available on the usual 192.168.1.1 address after a while.

If anything is not OK with the patch please let me know so I can fix it.

Thanks and best regards.

araujorm commented:
Also, here's the forum thread where this device has been discussed and where the patch was tested with help from some other helpful forum members and openwrt users:

https://forum.openwrt.org/t/adding-support-for-comfast-cf-wr758ac-aka-joowin-jw-wr758ac/

mans0n commented:
Hi, please copy the PR description to the commit message.

And the common dts file should have ".dtsi" extension.

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac.dts:
```diff
@@ -0,0 +1,110 @@
+
+&pcie0 {
+	mt76@0,0 {
+		reg = <0x0000 0 0 0 0>;
```
This should be:
```diff
&pcie0 {
-	mt76@0,0 {
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
```

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac.dts:
```diff
@@ -0,0 +1,110 @@
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
```
Can it go faster? 40~50MHz would be safe in most cases.

araujorm replied:
Changed to 50000000 in latest push. 

> Hi, please copy the PR description to the commit message.
> 
> And the common dts file should have ".dtsi" extension.

Fixed in latest push.

araujorm replied:
Changed to 50000000 in latest push. Will still test it, so not marking this as resolved yet.

araujorm replied:
Has been working fine for the past 24 hours with 50MHz, value used in the latest push. Marking this as resolved.

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac.dts:
```diff
@@ -0,0 +1,110 @@
+			partition@50000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x50000 0x760000>;
```
What is after 0x7b0000?

araujorm replied:
This is as it is on the vendor firmware (see https://forum.openwrt.org/t/adding-support-for-comfast-cf-wr758ac-aka-joowin-jw-wr758ac/87355/21 for full dmesg output, as it's also possible to modify the vendor firmware and re-flash it in order to obtain root ssh access).

From vendor dmesg, the part that interests is:
```
[    0.690000] Creating 5 MTD partitions on "spi32766.0":
[    0.700000] 0x000000000000-0x000000030000 : "u-boot"
[    0.710000] 0x000000030000-0x000000040000 : "u-boot-env"
[    0.710000] 0x000000040000-0x000000050000 : "factory"
[    0.720000] 0x000000050000-0x0000007b0000 : "firmware"
```

And the /proc/mtd in vendor firmware after first boot:
```
root@JOOWIN:~# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00030000 00010000 "u-boot"
mtd1: 00010000 00010000 "u-boot-env"
mtd2: 00010000 00010000 "factory"
mtd3: 00760000 00010000 "firmware"
mtd4: 0012386d 00010000 "kernel"
mtd5: 0063c793 00010000 "rootfs"
mtd6: 00110000 00010000 "rootfs_data"
mtd7: 00040000 00010000 "configs"
```

What are your recommendations?

mans0n replied:
If the flash memory area after 0x7b0000 is indeed empty, then you might be able to extend the firmware partition.
If not (or to be safe), keep it as is. In that case you can add a note that 0x7b0000~0x800000 is intentionally left unused.

araujorm replied:
Fixed in latest push, firmware partition now goes up to 0x800000. Tested with success.

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac.dts:
```diff
@@ -0,0 +1,110 @@
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0xe000>;
```
Please replace mtd-mac-address with nvmem-cells. (refer the other dts files)

araujorm replied:
I *think* I made it correctly in my latest push, can you please confirm?

araujorm replied:
For the record, the image compiled using the latest push (cherry-picked to v21.02.1), using nvmem-cells as I understood it should be, works as expected and the ethernet mac address is the same as in the vendor firmware (for the wifi mac address I didn't need to add anything, but it is also correct). So I really think I got it right in the last push but please confirm syntax, etc. Thanks.

mans0n replied:
LGTM at a glance.

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac_v1.dts:
```diff
@@ -0,0 +1,8 @@
+#include "mt7628an_joowin_jw-wr758ac.dts"
+
+/ {
+	compatible = "joowin,jw-wr758ac,v1", "joowin,jw-wr758ac", "mediatek,mt7628an-soc";
```
`"joowin,jw-wr758ac-v1"`

adschm replied:
And probably the second entry "joowin,jw-wr758ac" is useless and may be dropped (same below).

araujorm replied:
Fixed in latest push.

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac_v2.dts:
```diff
@@ -0,0 +1,8 @@
+#include "mt7628an_joowin_jw-wr758ac.dts"
+
+/ {
+	compatible = "joowin,jw-wr758ac,v2", "joowin,jw-wr758ac", "mediatek,mt7628an-soc";
```
Ditto.

mans0n commented on target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:
```diff
@@ -101,6 +101,10 @@ ramips_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1" "0:wan" "6@eth0"
 		;;
+	joowin,jw-wr758ac,v[12])
```
Ditto. And please separate this into two lines.

araujorm replied:
Fixed in latest push.

mans0n commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -220,6 +220,26 @@ define Device/iptime_a604m
+  DEVICE_MODEL := WR758AC
+endef
+
+define Device/joowin_jw-wr758ac_v1
```
`joowin_jw-wr758ac-v1`

araujorm replied:
Fixed in latest push.

araujorm commented:
> Hi, please copy the PR description to the commit message.
> 
> And the common dts file should have ".dtsi" extension.

Fixed in latest push.

araujorm commented:
New push: target/linux/ramips/mt76x8/base-files/etc/board.d/02_network was still referencing jw-wr758ac,v1 and jw-wr758ac,v2, should be jw-wr758ac-v1 and jw-wr758ac-v1.

araujorm commented:
Another new push: correct `mt7628an_joowin_jw-wr758ac_v*.dts` -> `target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac-v*.dts`

araujorm commented:
Should anyone be interested in pre-compiled images for both devices of v21.02.1 using this pull request and the official release config (thus having the same kernel vermagic and 100% compatible with official openwrt repos) please see the aforementioned thread, more specifically this post:

https://forum.openwrt.org/t/adding-support-for-comfast-cf-wr758ac-aka-joowin-jw-wr758ac/87355/64

(hopefully support can be added officially in one of the next releases of v21.02?)

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac.dtsi:
```diff
@@ -0,0 +1,122 @@
+
+&state_default {
+	gpio {
+		groups = "i2c", "wled_an", "gpio", "refclk", "wdt", "p0led_an";
```
Do you need i2c and p0led_an? It seems this would be enough:
```diff
 &state_default {
 	gpio {
-		groups = "i2c", "wled_an", "gpio", "refclk", "wdt", "p0led_an";
+		groups = "refclk", "wdt", "wled_an";
 		function = "gpio";
 	};
 };
```

araujorm replied:
Thanks, fixed in latest push.

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac.dtsi:
```diff
@@ -0,0 +1,122 @@
+};
+
+&wmac {
+	status = "okay";
```
Wait, in the latest snapshot you must have needed this:
```diff
 &wmac {
 	status = "okay";
+
+	mediatek,mtd-eeprom = <&factory 0x0>;
 };
```
Have you really had build/run tests with the patch on master branch?

araujorm replied:
Fixed in latest push.

I confess that at the time I was just testing with stable branch and didn't always test the cherry-picked commit on master branch. But this time I tested, the build of master branch succeeded and the device was in working state after reboot.

mans0n commented on target/linux/ramips/dts/mt7628an_joowin_jw-wr758ac.dtsi:
```diff
@@ -0,0 +1,122 @@
+};
+
+&esw {
+	mediatek,portmap = <0x2f>;
```
Do you need this line? This line means that port 4 is for wan, contrary to what you described in `02_network`.

araujorm replied:
Hi. Sorry for the late reply.

I wasn't sure what to put there, this was based on another dts I tried to make in the past for another device that had a different number of ports.

The joowin WR758AC devices only have one port, so the `02_network` file has the correct information. Should I just remove the `&esw` section then?

I've noticed meanwhile that there are already some other devices with the same switch definition:
```
        tplink,tl-wr902ac-v3|\
        wavlink,wl-wn576a2)
                ucidef_add_switch "switch0" \
                        "4:lan" "6@eth0"
                ;;
```
Should I move the joowin definitions close to those? Or, for alphabetical purposes, should I keep them as they are?

Thanks in advance.

araujorm replied:
Meanwhile, for example, the mt7628an_wavlink_wl-wn577a2 device also has `mediatek,portmap = <0x2f>;` so it must be also wrong?
Thanks in advice.

mans0n replied:
> The joowin WR758AC devices only have one port, so the `02_network` file has the correct information. Should I just remove the `&esw` section then?

My suggestion:
```
&esw {
	mediatek,portdisable = <0x2f>;
};
```
`mediatek,portdisable` is for turning off unused ports.

> Should I move the joowin definitions close to those? Or, for alphabetical purposes, should I keep them as they are?

It should be like this:
```diff
+	joowin,jw-wr758ac-v1|\
+	joowin,jw-wr758ac-v2|\
	tplink,tl-wr902ac-v3|\
	wavlink,wl-wn576a2)
		ucidef_add_switch "switch0" \
			"4:lan" "6@eth0"
		;;
```

> Meanwhile, for example, the mt7628an_wavlink_wl-wn577a2 device also has `mediatek,portmap = <0x2f>;` so it must be also wrong? Thanks in advice.

Probably.

araujorm replied:
Fixed in latest push.

mans0n commented:
Please provide more detailed hardware specification in the commit message. (Refer to 57cb387cfeee)
And the commit message should have at most 75 characters per line.

araujorm commented:
> Please provide more detailed hardware specification in the commit message. (Refer to [3c97fb4](https://github.com/openwrt/openwrt/commit/3c97fb4346d11ab3e7cc453ffd7b66b4b9140aec)) And the commit message should have at most 75 characters per line.

I have some doubts:
- this device does not require an external power supply; should I just ommit the Power entry?
- I can see that the "MAC addresses map" section is related to the offsets on the factory partition, but I'm not sure what `:3e` and `*:3c` mean. Can you help?

Best regards.

araujorm commented:
> * I can see that the "MAC addresses map" section is related to the offsets on the factory partition, but I'm not sure what `:3e` and `*:3c` mean. Can you help?

If it helps, all the MAC addresses of both devices start with 20:0D:B0, I would understand if there was something like `20:0D:B0:*`... is that what's intended?

araujorm committed:
ramips: add support for Joowin WR758AC V1 and V2

This commit adds support for Joowin (aka Comfast) WR758AC V1 and V2
devices.

Both have the same wall AP/repeater form factor and differ only
in the 5Ghz chipset (V1 has MT7662, V2 has MT7663).

OpenWrt developers forum page:
https://forum.openwrt.org/t/87355

Specifications:

 - CPU: MediaTek MT7628AN (580MHz)
 - Flash: 8MB
 - RAM: 64MB DDR2
 - 2.4 GHz: 802.11b/g/n (MT7603)
 - 5 GHz: 802.11ac (V1 has MT7662, V2 has MT7663)
 - Antennas: 4x external single band antennas
 - LAN: 1x 10/100M
 - LED: Wifi 3x blue. Programmable
 - Button: WPS

MAC addresses as verified by OEM firmware:

use address source
LAN *:83 factory 0xe000
2g *:85 factory 0x4
5g *:86 factory 0x8004

How to install:
1- Setup a TFTP server on a machine with IP address 192.168.1.10/24
2- Name the image as `firmware_auto.bin` and place it on the root of the
TFTP server
3- Connect the device via Ethernet, it should pick and flash the image

Signed-off-by: Rodrigo Araujo <araujo.rm@gmail.com>

mans0n commented:
> * this device does not require an external power supply; should I just ommit the Power entry?

The vendor webpage says the device is a wireless repeater, then it is powered by AC outlet, isn't it?

> * I can see that the "MAC addresses map" section is related to the offsets on the factory partition, but I'm not sure what `:3e` and `*:3c` mean. Can you help?

It means that the MAC address ends with `3e` or `3c`. In most cases the MAC addresses of a device differ only in the last byte.

n-kremeris commented:
I'm testing the latest FW from the link in the forum posts (i've a v2 unit) and I'm happy to report that everything works well so far - will give an update if i encounter any problems.

araujorm commented:
> Please provide more detailed hardware specification in the commit message. (Refer to [3c97fb4](https://github.com/openwrt/openwrt/commit/3c97fb4346d11ab3e7cc453ffd7b66b4b9140aec)) And the commit message should have at most 75 characters per line.

Please validate if the latest message is as intended. Thanks.

hauke commented:
Thank you for the patch, I applied it to master in 766733e17226221a310a30c91927a79f07a456e5.

mans0n commented:
Sorry, has been busy for a while.

Based on the forum thread and googling and `git grep`, I suppose the original vendor is Comfast, and Joowin and WJOY models are rebrandings of Comfast one. If my guess is correct, then I would prefer "Comfast CF-WR758AC" as the primary model name.
@araujorm Could you consider renaming the model name (and moving "Joowin JW-WR758AC" to `DEVICE_ALT0_*`) in another patch?

araujorm commented:
> 

Sure. Something like this? https://github.com/araujorm/openwrt/commit/ff7c6647a0bdbba95b4c7a773c4920ec39f49a09
(I haven't tested building it, but at least on make menuconfig it seems to work, if you say so I'll proceed with building and testing and later create a PR with this)

mans0n commented:
Looks good to me.
I'll leave some other comments after you open the PR.

araujorm commented:
Ok, compiles without errors here. Device boots, wifi and ethernet work.

So I closed the previous PR and opened a new one that fixes the naming and indirectly fixes the sorting:
https://github.com/openwrt/openwrt/pull/4961
